### ISSUE TITLE
Output resources and their configurations into a json file before deployment for further inspection

### ISSUE THREAD

**Issue by XuejiaoZhang** [pulumi/pulumi#7906]
## Policy-as-Code check against Pulumi Infra-as-Code

We are planning to use OPA to run compliance check against the resources to be deployed via Pulumi.
OPA will take any json input file to compare the configurations of resources to the desired status.
For Terraform, with `Terraform plan` we could get the output json and integrate with OPA, please see more info [here](https://www.openpolicyagent.org/docs/latest/terraform/)
For CloudFormation, it has a json template which we could run OPA policies against it, please see [here](https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/gettingstarted.templatebasics.html#gettingstarted.templatebasics.what).
However, for Pulumi, I didn't see any output json similar which contains all the resources and their configurations. Even the output of `pulumi preview -j`, it is a bit troublesome if we use it to run OPA policy directly against.

I am wondering is there a way to have Pulumi output a json file with all the resources and their configurations, so that we could run OPA policies directly against it, instead of using [Pulumi CrossGuard to run OPA](https://github.com/pulumi/pulumi-policy-opa) (because of [this](https://github.com/pulumi/pulumi-policy-opa/issues/9)).

**Comment by emiliza** [comment]
Have you tried `pulumi stack export`? https://www.pulumi.com/docs/intro/concepts/stack/#import-and-export-a-stack-deployment

**Comment by XuejiaoZhang** [comment]
1. I would like to have the output with all the resources and their configurations automatically, similar to [this CloudFormation template](https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/quickref-ec2.html)
For Pulumi, It seems I need to add `pulumi.export()` in the policy code, in order to get all the resources, I need to add `export` for all the resources in the code.

2. If I understand it correctly, `pulumi stack export` will output only after running `Pulumi up`, that mean  all the resources have been created. We are planning to do pre-check before deployment so that if it violate our policies, we can fail the deployment.

**Comment by lukehoban** [comment]
The feature being implemented to support #2318 will provide a way to get a plan file as an output of `pulumi preview` which can be used for this purpose.

Note that for you original use case, you could also use Pulumi Policy, which is a policy engine built into the Pulumi CLI/engine. It supports OPA as an authoring format. 

I’ll close this out as a duplicate of #2318 given that that feature will enable this use case.
